### PR TITLE
fix(failover): add failover timeout to FailoverDomain

### PR DESCRIPTION
**Pull request by fimanishi**
**What changed?**
Added FailoverTimeoutInSeconds to FailoverDomainRequest, wired it for thrift, proto, and cli, and added validation setting minimum and maximum values (using dynamic config).

Also allows the endpoint FailoverDomain to perform graceful failover, added metrics and tests.

https://github.com/cadence-workflow/cadence/issues/8139

**Why?**
When the FailoverDomainRequest was created, FailoverTimeoutInSeconds was not added to it. This caused the FailoverDomain api to not accept graceful failovers.

**How did you test it?**
go test -run 'TestValidateFailoverDomainRequest' ./service/frontend/api/...
go test -run 'TestFailoverDomainRequestConversion' ./common/types/mapper/thrift/...
go test -run 'TestFailoverDomainRequest|TestFailoverDomainRequestFuzz' ./common/types/mapper/proto/...
go test -race -run TestFailoverDomain ./service/frontend/api/...

**Potential risks**
There are no major risks. This is just wiring graceful failover to the new failover endpoint.

**Release notes**
Add graceful failover option through FailoverTimeoutInSeconds to the Failover api

**Documentation Changes**
N/A

### CODE REVIEW

File: tools/cli/domain_utils.go [hunk 1]
```diff
@@ -290,6 +290,11 @@ var (
 			Aliases: []string{"r"},
 			Usage:   "Reason for failover (for tracking and transparency)",
 		},
+		&cli.IntFlag{
+			Name:    FlagFailoverTimeout,
+			Aliases: []string{"fts"},
+			Usage:   "[Optional] Graceful failover timeout in seconds. If set, the failover is performed gracefully instead of force.",
```
Comment (c-warren) [inline tools/cli/domain_utils.go]: I think we should say something like 'if set, the failover will happen after all tasks have drained from the active cluster, or this timeout is reached - whichever is earlier'. I'm not sure that just saying gracefully/forced helps the user know the repercussions. wdyt?

File: service/frontend/api/request_validator_test.go [hunk 2]
```diff
@@ -1102,3 +1102,127 @@ func TestValidateRegisterDomainRequest(t *testing.T) {
...
+func TestValidateFailoverDomainRequest(t *testing.T) {
+	testCases := []struct {
+		name          string
+		req           *types.FailoverDomainRequest
+		lockdown      bool
+		expectError   bool
+		expectedError string
+	}{
+		{
+			name: "success - active cluster",
```
Comment (c-warren) [inline service/frontend/api/request_validator_test.go]: NIT: I prefer writing test case names with a clear sentence describing when X happens it should result Z (https://github.com/recht/rants/blob/master/testing.md#proper-test-naming). 

This helps a lot when debugging the test later - you have a clear understanding of what was supposed to happen, why it was supposed to happen + what actually happened (from the test output).

File: common/types/shared.go [hunk 3]
```diff
@@ -1829,6 +1831,14 @@ func (v *FailoverDomainRequest) GetReason() (o string) {
 	return
 }
 
+// GetFailoverTimeoutInSeconds is an internal getter (TBD...)
```
Comment (c-warren) [inline common/types/shared.go]: I know we have this comment everywhere. Imo just remove it if we don't have anything to say, or add documentation.

File: common/types/mapper/thrift/shared_test.go [hunk 4]
```diff
@@ -3716,3 +3716,14 @@ func TestListFailoverHistoryResponseConversion(t *testing.T) {
...
+func TestFailoverDomainRequestConversion(t *testing.T) {
+	for _, item := range []*types.FailoverDomainRequest{
+		nil,
+		{},
+		&testdata.FailoverDomainRequest,
+		&testdata.FailoverDomainRequest_OnlyActiveClusters,
+	} {
+		assert.Equal(t, item, ToFailoverDomainRequest(FromFailoverDomainRequest(item)))
+	}
+}
```
Comment (c-warren) [inline common/types/mapper/thrift/shared_test.go:3729]: I think we should have a fuzz test as well for this, to ensure that we test edge cases for all the fields here. We can probably skip ActiveClusters as we have a lot of tests already testing that (so the sanity check is probably enough here). 

Unless we already do have one in the proto/api_test.go?

File: service/frontend/api/domain_handlers.go [hunk 5]
```diff
@@ -105,7 +106,21 @@ func (wh *WorkflowHandler) UpdateDomain(
...
+	emitInitiationFailure := func(reason string) {
+		if !isGraceFailover {
+			return
+		}
+		wh.GetMetricsClient().Scope(
+			metrics.FrontendUpdateDomainScope,
+			metrics.DomainTag(updateRequest.GetName()),
+			metrics.ReasonTag(reason),
+		).IncCounter(metrics.GracefulFailoverInitiationFailure)
+	}
```
Comment (c-warren) [inline service/frontend/api/domain_handlers.go]: I'd just extract this out into a function, we're using it in update domain and failover domain.

**Review by c-warren — APPROVED** [review]
